### PR TITLE
Pathlibify animation.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/animation.py [hunk 1]
```diff
@@ -826,10 +824,10 @@ def setup(self, fig, outfile, dpi, frame_dir=None):
 
         if not self.embed_frames:
             if frame_dir is None:
-                frame_dir = root + '_frames'
-            if not os.path.exists(frame_dir):
-                os.makedirs(frame_dir)
-            frame_prefix = os.path.join(frame_dir, 'frame')
+                frame_dir = Path(str(outfile.with_suffix('')) + '_frames')
+            if not frame_dir.exists():
+                frame_dir.mkdir(parents=True)
```
Comment (QuLogic) [inline lib/matplotlib/animation.py]:
```suggestion
            frame_dir.mkdir(parents=True, exist_ok=True)
```

Comment (eric-wieser) [reply lib/matplotlib/animation.py]: Except replacing both lines, not just the second one

File: lib/matplotlib/animation.py [hunk 2]
```diff
@@ -826,10 +824,10 @@ def setup(self, fig, outfile, dpi, frame_dir=None):
...
+            frame_prefix = str(frame_dir / 'frame')
```
Comment (eric-wieser) [inline lib/matplotlib/animation.py]: Why coerce to string here?

File: lib/matplotlib/animation.py [hunk 3]
```diff
@@ -826,10 +824,10 @@ def setup(self, fig, outfile, dpi, frame_dir=None):
 
         if not self.embed_frames:
             if frame_dir is None:
-                frame_dir = root + '_frames'
-            if not os.path.exists(frame_dir):
-                os.makedirs(frame_dir)
-            frame_prefix = os.path.join(frame_dir, 'frame')
+                frame_dir = Path(str(outfile.with_suffix('')) + '_frames')
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/animation.py]: No, that would drop outfile's directory.  I guess `outfile.parent / (outfile.stem + '_frames')` is nicer, though.

File: lib/matplotlib/animation.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/animation.py]: changed

File: lib/matplotlib/animation.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/animation.py]: I was worried the super() method would not handle Path objects, but it looks like it does.

File: lib/matplotlib/animation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/matplotlib/animation.py]: What do you mean?
```python3
In [3]: outfile = Path('a/b/c'); outfile.with_name(outfile.stem + '_frames')
Out[3]: PosixPath('a/b/c_frames')

In [4]: outfile = Path('a/b/c.mp4'); outfile.with_name(outfile.stem + '_frames')
Out[4]: PosixPath('a/b/c_frames')

In [5]: outfile = Path('/a/b/c.mp4'); outfile.with_name(outfile.stem + '_frames')
Out[5]: PosixPath('/a/b/c_frames')

```

[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/animation.py]: oops, sorry, misread.  fixed accordingly.

**Comment by dstansby** [comment]
Anyone know what's happened to codecov? The changes look fine, but I'd be nervous approving lines that don't have code coverage.

**Comment by anntzer** [comment]
Dunno what happened with codecov, but you can run the tests locally and check coverage yourself (`pytest lib/matplotlib/tests/test_animation.py --cov && coverage html`); all lines touched by this PR are covered, except that in
```patch
-        root, ext = os.path.splitext(outfile)
-        if ext not in ['.html', '.htm']:
+        outfile = Path(outfile)
+        if outfile.suffix not in ['.html', '.htm']:
```
(which raises an exception when the suffix is wrong)
the throwing case was never exercised, and is still not.

**Comment by tacaswell** [comment]
Thank you @anntzer !

**Pull request merged**